File: unity/simple-stack-vm-unity/Assets/Scripts/SimpleStackVM/StandardLibrary/StandardObjectLibrary.cs
using System;
using System.Collections.Generic;
using System.Linq;

#nullable enable

namespace SimpleStackVM
{
    public static class StandardObjectLibrary
    {
        #region Fields
        public static readonly IReadOnlyScope Scope = CreateScope();
        #endregion

        #region Methods
        public static Scope CreateScope()
        {
            var result = new Scope();

            var objectFunctions = new Dictionary<string, IValue>
            {
                {"set", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var value = vm.PopStack();
                    var key = vm.PopStack<StringValue>();
                    var obj = vm.PopStack<ObjectValue>();
                    vm.PushStack(Set(obj, key.Value, value));
                })},

                {"get", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var key = vm.PopStack<StringValue>();
                    var obj = vm.PopStack<ObjectValue>();
                    if (obj.TryGetValue(key.Value, out var value))
                    {
                        vm.PushStack(value);
                    }
                    else
                    {
                        vm.PushStack(NullValue.Value);
                    }
                })},

                {"removeKey", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var key = vm.PopStack<StringValue>();
                    var obj = vm.PopStack<ObjectValue>();
                    vm.PushStack(RemoveKey(obj, key.Value));
                })},

                {"removeValues", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var values = vm.PopStack();
                    var obj = vm.PopStack<ObjectValue>();
                    vm.PushStack(RemoveValues(obj, values));
                })},

                {"keys", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var top = vm.PopStack<ObjectValue>();
                    vm.PushStack(Keys(top));
                })},

                {"values", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var top = vm.PopStack<ObjectValue>();
                    vm.PushStack(Values(top));
                })},

                {"length", new BuiltinFunctionValue((vm, numArgs) =>
                {
                    var top = vm.PopStack<ObjectValue>();
                    vm.PushStack(top.Value.Count);
                })}
            };

            result.Define("object", new ObjectValue(objectFunctions));

            return result;
        }

        public static ArrayValue Keys(ObjectValue self)
        {
            var keys = self.Value.Keys.Select(k => new StringValue(k) as IValue).ToList();
            return new ArrayValue(keys);
        }

        public static ArrayValue Values(ObjectValue self)
        {
            return new ArrayValue(self.Value.Values.ToList());
        }

        public static ObjectValue Set(ObjectValue self, string key, IValue value)
        {
            var result = new Dictionary<string, IValue>(self.Value);
            result[key] = value;
            return new ObjectValue(result);
        }

        public static ObjectValue RemoveKey(ObjectValue self, string key)
        {
            if (!self.Value.ContainsKey(key))
            {
                return self;
            }

            var result = new Dictionary<string, IValue>(self.Value.Where(kvp => kvp.Key != key));
            return new ObjectValue(result);
        }

        public static ObjectValue RemoveValues(ObjectValue self, IValue values)
        {
            var result = new Dictionary<string, IValue>(self.Value.Where(kvp => kvp.Value.CompareTo(values) != 0));
            return new ObjectValue(result);
        }

        public static int GeneralCompareTo(IObjectValue left, IValue? rightInput)
        {
            if (left == rightInput)
            {
                return 0;
            }

            if (left == null || !(rightInput is IObjectValue right))
            {
                return 1;
            }

            var leftKeys = left.ObjectKeys;
            var rightKeys = right.ObjectKeys;
            var compareLength = leftKeys.Count.CompareTo(rightKeys.Count);
            if (compareLength != 0)
            {
                return compareLength;
            }

            foreach (var key in leftKeys)
            {
                if (!left.TryGetValue(key, out var leftValue))
                {
                    throw new Exception($"Object key: {key} not present in the object!: {left.ToString()}");
                }

                if (right.TryGetValue(key, out var rightValue))
                {
                    var compare = leftValue.CompareTo(rightValue);
                    if (compare != 0)
                    {
                        return compare;
                    }
                }
                else
                {
                    return 1;
                }
            }

            return 0;
        }
        #endregion
    }
}

File: ts/src/standardLibrary/standardObjectLibrary.ts
import Scope, { IReadOnlyScope } from "../scope";
import { Editable } from "../types";
import ArrayValue from "../values/arrayValue";
import BuiltinFunctionValue from "../values/builtinFunctionValue";
import { IValue } from "../values/ivalues";
import NullValue from "../values/nullValue";
import ObjectValue, { isObjectValue, ObjectValueMap } from "../values/objectValue";
import StringValue from "../values/stringValue";

export const objectScope: IReadOnlyScope = createObjectScope();

export function createObjectScope()
{
    const result = new Scope();

    const objectFunctions: ObjectValueMap =
    {
        set: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getCast(0, isObjectValue);
            const key = args.getString(1);
            const value = args.get(2);
            vm.pushStack(set(top, key, value));
        }),

        get: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getCast(0, isObjectValue);
            const key = args.getString(1);
            const found = get(top, key);
            if (found !== undefined)
            {
                vm.pushStack(found);
            }
            else
            {
                vm.pushStack(NullValue.Value);
            }
        }),

        removeKey: new BuiltinFunctionValue((vm, args) =>
        {
            const obj = args.getCast(0, isObjectValue);
            const key = args.getString(1);
            vm.pushStack(removeKey(obj, key));
        }),

        removeValues: new BuiltinFunctionValue((vm, args) =>
        {
            const obj = args.getCast(0, isObjectValue);
            const values = args.get(1);
            vm.pushStack(removeValues(obj, values));
        }),

        keys: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getCast(0, isObjectValue);
            vm.pushStack(keys(top));
        }),

        values: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getCast(0, isObjectValue);
            vm.pushStack(values(top));
        }),

        length: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getCast(0, isObjectValue);
            vm.pushStackNumber(length(top));
        })
    }

    result.define('object', new ObjectValue(objectFunctions));

    return result;
}

export function set(target: ObjectValue, key: string, value: IValue)
{
    return new ObjectValue({ ...target.value, [key]: value });
}

export function get(target: ObjectValue, key: string) : IValue | undefined
{
    const result = target.value[key];
    return result != null ? result : undefined;
}

export function keys(target: ObjectValue): ArrayValue
{
    return new ArrayValue(target.keys.map(s => new StringValue(s)));
}

export function removeKey(target: ObjectValue, key: string): ObjectValue
{
    if (!target.hasOwnProperty(key))
    {
        return target;
    }

    const result = { ...target.value };
    delete result[key];
    return new ObjectValue(result);
}

export function removeValues(target: ObjectValue, values: IValue): ObjectValue
{
    const result: Editable<ObjectValueMap> = {};
    for (const key in target)
    {
        if (target.value[key].compareTo(values) !== 0)
        {
            result[key] = target.value[key];
        }
    }

    return new ObjectValue(result);
}

export function values(target: ObjectValue): ArrayValue
{
    return new ArrayValue(Object.values(target.value));
}

export function length(target: ObjectValue)
{
    return target.keys.length;
}